src/weapons.cpp

```diff
@@ -0,0 +1,1526 @@
+            Audio_Send(weapon.shot, TR_AUDIO_EMITTER_ENTITY, ent->id);
+            consumeAmmo(weapon);
+
+            {
+                collision_result_t cs;
+                float from[3], to[3], tr[16];
+                ss_bone_tag_p bt = ent->bf->bone_tags + targeted_bone_start;
+
+                Mat4_Mat4_mul(tr, ent->transform.M4x4, bt->current_transform);
+                Mat4_vec3_mul(from, ent->transform.M4x4, ent->bf->bone_tags[targeted_bone_end].current_transform + 12);
+
+                if (target && (bt->mod.current_slerp > 0.99))
+                {
+                    vec3_copy(to, bt->mod.target_pos);
+                }
+                else
+                {
+                    vec3_add_mul(to, from, tr + 8, -32768.0f);
+                }
+
+                if (Physics_RayTest(&cs, from, to, ent->self, COLLISION_FILTER_CHARACTER) && cs.obj && (cs.obj->object_type == OBJECT_ENTITY))
+                {
+                    target = (entity_p)cs.obj->object;
+                    Script_ExecEntity(engine_lua, ENTITY_CALLBACK_SHOOT, ent->id, target->id);
+                    return true;
+                }
+            }
+        }
+        else
+        {
+            Anim_SetAnimation(ss_anim, anim_firing_to_idle, -1);
+            return true;
+        }
+    }
+
+    return false;
+}
+
+#pragma region weapons
+struct weapons_s getPistol()
+{
+    return pistol;
+}
+
+struct weapons_s getShotgun()
+{
+    return shotgun;
+}
+
+struct weapons_s getMagnum()
+{
+    return magnum;
+}
+
+struct weapons_s getUzi()
+{
+    return uzi;
+}
+
+struct weapons_s getAutomags()
+{
+    return automags;
+}
+
+struct weapons_s getDesertEagle()
+{
+    return deserteagle;
+}
+
+struct weapons_s getRevolver()
+{
+    return revolver;
+}
+
+struct weapons_s getM16()
+{
+    return m16;
+}
+
+struct weapons_s getMP5()
+{
+    return mp5;
+}
+
+struct weapons_s getRocketGun()
+{
+    return rocketgun;
+}
+
+struct weapons_s getHarpoonGun()
+{
+    return harpoongun;
+}
+
+struct weapons_s getGrenadeGun()
+{
+    return grenadegun;
+}
+
+struct weapons_s getCrossbowGun()
+{
+    return crossbowgun;
+}
+
+struct weapons_s getGrapplinGun()
+{
+    return grapplingun;
+}
+
+/* 
+ * - used to get all weapon in one function
+ *- 0: pistol;
+ *- 1: shotgun;
+ *- 2: magnum;
+ *- 3: uzi;
+ *- 4: automags;
+ *- 5: desert_eagle;
+ *- 6: revolver;
+ *- 7: m16;
+ *- 8: mp5;
+ *- 9: rocketgun;
+ *- 10: harpoongun;
+ *- 11: grenadegun; 
+ *- 12: crossbowgun;
+ *- 13: grapplingun;
+ */
+struct weapons_s *getAll()
+{
+    weapons_s weapons[14];
+    weapons[0] = pistol;
+    weapons[1] = shotgun;
+    weapons[2] = magnum;
+    weapons[3] = uzi;
+    weapons[4] = automags;
+    weapons[5] = deserteagle;
+    weapons[6] = revolver;
+    weapons[7] = m16;
+    weapons[8] = mp5;
+    weapons[9] = rocketgun;
+    weapons[10] = harpoongun;
+    weapons[11] = grenadegun;
+    weapons[12] = crossbowgun;
+    weapons[13] = grapplingun;
+    return weapons;

```

Comment:
Be careful, `weapons` is a local variable. This function won't work if called.

Reply:
#delete_and_fixed